src/stactools/viirs/cog.py

```diff
@@ -0,0 +1,158 @@
+
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore", category=NotGeoreferencedWarning)
+            with rasterio.open(rasterio_subdataset_path, "r") as tag_src:
+                rasterio_tags = tag_src.tags()
+
+        with h5py.File(infile) as h5:
+            data: Any = np.array(h5[subdataset_key])
+            if len(data.shape) == 1:  # skip single value (non-data) "grids"
+                continue
+            if len(data.shape) == 3:
+                raise ValueError(
+                    f"MultiBand COG creation not supported for "
+                    f"subdataset '{subdataset_name}' in product "
+                    f"'{metadata.product}'"
+                )
+
+            # gdal (and software built on gdal) doesn't always play well with signed byte data
+            data = np.int16(data) if data.dtype == "int8" else data
+
+            if "_FillValue" in h5[subdataset_key].attrs.keys():
+                temp = h5[subdataset_key].attrs["_FillValue"].item()
+            elif "_Fillvalue" in h5[subdataset_key].attrs.keys():
+                temp = h5[subdataset_key].attrs["_Fillvalue"].item()
+            nodata_values: List[Optional[int]] = [None] if temp == b"n/a" else [temp]
+
+            nodata_new: int
+            if metadata.product in MULTIPLE_NODATA:
+                subdatasets = MULTIPLE_NODATA[metadata.product]
+                if subdataset_name in subdatasets:
+                    nodata_values = cast(
+                        List[Optional[int]], subdatasets[subdataset_name]["multiple"]
+                    )
+                    nodata_new = cast(int, subdatasets[subdataset_name]["new"])
+
+            if len(nodata_values) > 1:
+                clean_data, clean_nodata = clean(data, nodata_values, nodata_new)
+                nodata_cog_path = f"{os.path.splitext(cog_path)[0]}_fill.tif"
+                cog(
+                    clean_data,
+                    metadata.crs,
+                    metadata.transform,
+                    rasterio_tags,
+                    cog_path,
+                    nodata_new,
+                )
+                cog_paths.append(cog_path)
+                cog(
+                    clean_nodata,
+                    metadata.crs,
+                    metadata.transform,
+                    rasterio_tags,
+                    nodata_cog_path,
+                    nodata_new,
+                )
+                cog_paths.append(nodata_cog_path)
+            else:
+                cog(
+                    data,
+                    metadata.crs,
+                    metadata.transform,
+                    rasterio_tags,
+                    cog_path,
+                    nodata_values[0],
+                )
+                cog_paths.append(cog_path)
+
+    return cog_paths
+
+
+def cog(
+    data: Any,
+    crs: str,
+    transform: List[float],
+    tags: Dict[str, Any],
+    cog_path: str,
+    nodata: Optional[int] = None,
+) -> None:
```

Comment:
This function should either be documented or made "private" (rename to `_cog`).

Reply:
Yep. Done. I split these code chunks out for dry only